File: CSharp/refClient.vb
Imports System.Collections.Concurrent
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Threading.Tasks
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq
Imports System.Net


Public Class refClient
	Dim BaseUrl As String = "https://pubquick.co.uk/"

	Public Function Parse(orefquery As Query) As QueryResult
		Try
			Using wclient As New Net.Http.HttpClient()
				wclient.BaseAddress = New Uri(BaseUrl)
				wclient.DefaultRequestHeaders.Accept.Clear()
				wclient.DefaultRequestHeaders.Accept.Add(New Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"))
				wclient.Timeout = TimeSpan.FromMinutes(30)

				Dim requestcontent = New StringContent(JsonConvert.SerializeObject(orefquery), System.Text.Encoding.UTF8, "application/json")
				'Dim postTask = wclient.PostAsJsonAsync("api/refparser", orefquery)
				Dim postTask = wclient.PostAsync("api/refparser", requestcontent)
				postTask.Wait()
				Dim result = postTask.Result
				If result.IsSuccessStatusCode Then
					Dim readTask = result.Content.ReadAsStringAsync()
					readTask.Wait()
					Dim converters As Newtonsoft.Json.JsonConverter() = {New RefItemJsonConverter()}
					Dim qresult As QueryResult = Newtonsoft.Json.JsonConvert.DeserializeObject(Of QueryResult)(readTask.Result, New Newtonsoft.Json.JsonSerializerSettings() With {.Converters = converters})
					If qresult IsNot Nothing AndAlso qresult.Refs.Count > 0 Then
						Return qresult
					Else
						Return Nothing
					End If
				Else
					Return Nothing
				End If
			End Using
		Catch ex As Exception
			'ExceptionHandling(ex)
		End Try
		Return Nothing
	End Function
End Class

Public Class RefItemJsonConverter
	Inherits JsonConverter
	Public Overrides Function CanConvert(ByVal objectType As Type) As Boolean
		If objectType = GetType(refData.refitems) Then
			Return True
		End If
		Return False
	End Function

	Public Overrides Function ReadJson(ByVal reader As JsonReader, ByVal objectType As Type, ByVal existingValue As Object, ByVal serializer As JsonSerializer) As Object
		Dim jo As JObject = JObject.Load(reader)

		If jo("et") IsNot Nothing Then
			Select Case jo("et").Value(Of String)
				Case "grp" : Return jo.ToObject(Of refData.group)(serializer)
				Case "ri" : Return jo.ToObject(Of refData.refitem)(serializer)
				Case "sp" : Return jo.ToObject(Of refData.space)(serializer)
				Case Else
			End Select
		Else
			Return jo.ToObject(Of refData.refitem)(serializer)
		End If

		Return Nothing
	End Function

	Public Overrides ReadOnly Property CanWrite As Boolean
		Get
			Return False
		End Get
	End Property

	Public Overrides Sub WriteJson(ByVal writer As JsonWriter, ByVal value As Object, ByVal serializer As JsonSerializer)
		Throw New NotImplementedException()
	End Sub
End Class




Public Class QueryResult
	Public Property Result As Boolean = False
	Public Property Remarks As String = String.Empty
	Public Property timetaken As Integer
	Public Property Refs As New List(Of refData)
End Class

Public Class refData
	Public refmeta As New crefinfo
	Public refs As New List(Of refitems)

	Public Property result As String
	Public Property matchpercent As Integer = 0
	Public Property id As String = String.Empty
	Public Property rid As Long = 0
	Public Property seq As Integer = 0
	Public Property timetaken As Integer = 0
	Public Property reftype As String = String.Empty
	Public Class crefinfo
		Public Property refid As String = String.Empty
		Public Property doi As String = String.Empty
		Public Property pmid As String = String.Empty
		Public Property pmcid As String = String.Empty
		Public Property pii As String = String.Empty
		Public Property isbn As String = String.Empty
		Public Property pissn As String = String.Empty
		Public Property eissn As String = String.Empty
	End Class

	Public MustInherit Class refitems

	End Class
	Public Class group : Inherits refitems
		Public ReadOnly Property et As String = "grp"
		Public Property type As String = String.Empty
		Public refs As New List(Of refitems)
	End Class
	Public Class refitem : Inherits refitems
		Public ReadOnly Property et As String = "ri"
		Public Property type As String = String.Empty
		Public Property text As String = String.Empty
	End Class
	Public Class space : Inherits refitems
		Public ReadOnly Property et As String = "sp"
		Public Property type As String = "space"
		Public Property text As String = " "
	End Class
End Class
Public Class Query
	Public Enum epriority
		low = 1
		medium = 2
		high = 3
	End Enum
	Public Property webkey As String = String.Empty
	Public Property priority As epriority = epriority.medium
	Public Property batchname As String = String.Empty
	Public Property Demomode As Boolean = False
	Public Property refs As New List(Of refitem)


	Public Class refitem
		Public Property id As String = String.Empty
		Public Property seq As Integer = 0
		Public Property txt As String = String.Empty

		Public Property RefData As refData
	End Class
End Class

File: CSharp/refClient.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System.Collections.Concurrent;
using System.Net.Http;
using System.Net.Http.Headers;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Net;


public class refClient
{
    private string BaseUrl = "https://pubquick.co.uk/";

    public QueryResult Parse(Query orefquery)
    {
        try
        {
            using (System.Net.Http.HttpClient wclient = new System.Net.Http.HttpClient())
            {
                wclient.BaseAddress = new Uri(BaseUrl);
                wclient.DefaultRequestHeaders.Accept.Clear();
                wclient.DefaultRequestHeaders.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));
                wclient.Timeout = TimeSpan.FromMinutes(30);

                var requestcontent = new StringContent(JsonConvert.SerializeObject(orefquery), System.Text.Encoding.UTF8, "application/json");
                // Dim postTask = wclient.PostAsJsonAsync("api/refparser", orefquery)
                var postTask = wclient.PostAsync("api/refparser", requestcontent);
                postTask.Wait();
                var result = postTask.Result;
                if (result.IsSuccessStatusCode)
                {
                    var readTask = result.Content.ReadAsStringAsync();
                    readTask.Wait();
                    Newtonsoft.Json.JsonConverter[] converters = new[] { new RefItemJsonConverter() };
                    QueryResult qresult = Newtonsoft.Json.JsonConvert.DeserializeObject<QueryResult>(readTask.Result, new Newtonsoft.Json.JsonSerializerSettings() { Converters = converters });
                    if (qresult != null && qresult.Refs.Count > 0)
                        return qresult;
                    else
                        return null;
                }
                else
                    return null;
            }
        }
        catch (Exception ex)
        {
        }
        return null;
    }
}

public class RefItemJsonConverter : JsonConverter
{
    public override bool CanConvert(Type objectType)
    {
        if (objectType == typeof(refData.refitems))
            return true;
        return false;
    }

    public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
    {
        JObject jo = JObject.Load(reader);

        if (jo("et") != null)
        {
            switch (jo("et").Value<string>)
            {
                case "grp":
                    {
                        return jo.ToObject<refData.group>(serializer);
                    }

                case "ri":
                    {
                        return jo.ToObject<refData.refitem>(serializer);
                    }

                case "sp":
                    {
                        return jo.ToObject<refData.space>(serializer);
                    }

                default:
                    {
                        break;
                    }
            }
        }
        else
            return jo.ToObject<refData.refitem>(serializer);

        return null;
    }

    public override bool CanWrite
    {
        get
        {
            return false;
        }
    }

    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
    {
        throw new NotImplementedException();
    }
}




public class QueryResult
{
    public bool Result { get; set; } = false;
    public string Remarks { get; set; } = string.Empty;
    public int timetaken { get; set; }
    public List<refData> Refs { get; set; } = new List<refData>();
}

public class refData
{
    public crefinfo refmeta = new crefinfo();
    public List<refitems> refs = new List<refitems>();

    public string result { get; set; }
    public int matchpercent { get; set; } = 0;
    public string id { get; set; } = string.Empty;
    public long rid { get; set; } = 0;
    public int seq { get; set; } = 0;
    public int timetaken { get; set; } = 0;
    public string reftype { get; set; } = string.Empty;
    public class crefinfo
    {
        public string refid { get; set; } = string.Empty;
        public string doi { get; set; } = string.Empty;
        public string pmid { get; set; } = string.Empty;
        public string pmcid { get; set; } = string.Empty;
        public string pii { get; set; } = string.Empty;
        public string isbn { get; set; } = string.Empty;
        public string pissn { get; set; } = string.Empty;
        public string eissn { get; set; } = string.Empty;
    }

    public abstract class refitems
    {
    }
    public class group : refitems
    {
        public string et { get; } = "grp";
        public string type { get; set; } = string.Empty;
        public List<refitems> refs = new List<refitems>();
    }
    public class refitem : refitems
    {
        public string et { get; } = "ri";
        public string type { get; set; } = string.Empty;
        public string text { get; set; } = string.Empty;
    }
    public class space : refitems
    {
        public string et { get; } = "sp";
        public string type { get; set; } = "space";
        public string text { get; set; } = " ";
    }
}

public class Query
{
    public enum epriority
    {
        low = 1,
        medium = 2,
        high = 3
    }
    public string webkey { get; set; } = string.Empty;
    public epriority priority { get; set; } = epriority.medium;
    public string batchname { get; set; } = string.Empty;
    public bool Demomode { get; set; } = false;
    public List<refitem> refs { get; set; } = new List<refitem>();


    public class refitem
    {
        public string id { get; set; } = string.Empty;
        public long rid { get; set; } = 0;
        public int seq { get; set; } = 0;
        public string txt { get; set; } = string.Empty;

        public refData RefData { get; set; }
    }
}
